tools/opam-updates.py

```diff
@@ -0,0 +1,326 @@
+                        "(see https://github.com/settings/tokens). "
+                        "If omitted, the script will try to use the "
+                        "$HOME/.netrc file.")
+
+    parser.add_argument("--ignore-xs", dest="ignore_xs", action="store_true",
+                        help="Do not check if we are trying to "
+                        "upgrade packages in `xs` or `xs-extra`.")
+
+    args = parser.parse_args(argv)
+
+    if args.auth is None and args.xs_opam is not None:
+        sys.exit("Authentication is needed to be able to use the github api")
+
+    return args
+
+
+# Main body
+
+def main(argv=None):
+    """
+    Entry point.
+    """
+    args = parse_args_or_exit(argv)
+
+    # compute useful paths
+    if args.xs_opam is not None:
+        upstream_p = os.path.join(args.xs_opam, "packages", "upstream")
+        upstream_extra_p = os.path.join(
+            args.xs_opam, "packages", "upstream-extra")
+        xs_p = os.path.join(args.xs_opam, "packages", "xs")
+        xs_extra_p = os.path.join(args.xs_opam, "packages", "xs-extra")
+
+    out = get_opam_upgrade_output()
+
+    upgrades = [
+        extract_upgrade(row) for row in out
+        if "- upgrade" in row and "xenctrl" not in row
+    ]
+
+    updates = [
+        extract_update(row) for row in out
+        if "upstream changes" in row and "xenctrl" not in row
+    ]
+
+    new_packages = [
+        extract_install(row) for row in out
+        if "- install" in row and "xenctrl" not in row
+    ]
+
+    print_update_actions(upgrades, updates, new_packages)
+
+    if not args.xs_opam:
+        sys.exit(0)
+
+    # Passed xs-opam path as argument, perform update
+    upstream = set(
+        d for d in os.listdir(upstream_p)
+        if os.path.isdir(os.path.join(upstream_p, d)))
+
+    upstream_extra = set(
+        d for d in os.listdir(upstream_extra_p)
+        if os.path.isdir(os.path.join(upstream_extra_p, d)))
+
+    # Fail if we detect upstream updates for packages in xs or xs-extra
+    if not args.ignore_xs:
+        xs = set(
+            d for d in os.listdir(xs_p)
+            if os.path.isdir(os.path.join(xs_p, d))
+        ).union(
+            d for d in os.listdir(xs_extra_p)
+            if os.path.isdir(os.path.join(xs_extra_p, d))
+        )
+
+        for pkg in chain(upgrades, updates):
+            if pkg.current_name in xs:
+                sys.exit("Packages in `xs` or `xs-extra` should not be "
+                         "upgraded by opam: upgrading {}".format(
+                             pkg.current_name))
+
+    with requests.Session() as session:
+        session.headers.update({"Accept": "application/vnd.github.v3+json"})
+
+        for pkg in upgrades:
+            if pkg.current_name in upstream:
+                shutil.move(
+                    pkg.current_path(upstream_p),
+                    pkg.new_path(upstream_p))
+                update_files(session, args.auth, pkg, upstream_p)
```

Comment:
Will this remove files from the directory that have been deleted in the new version?

Reply:
There is a reason: if we had our own patches I was trying to avoid to delete them and leave them there, but I think you are right and it's better to just remove the folder and recreate it.

I'll make the change in a separate commit